File: 프로그래머스/3/68646. 풍선 터트리기/풍선 터트리기.cpp
#include <string>
#include <vector>
#include <cmath>
using namespace std;

int solution(vector<int> a) {
    if(a.size() == 1)
        return 1;
    
    int answer = 2;
    vector<int> leftMin (a.size(), 1000000001);
    vector<int> rightMin (a.size(), 1000000001);
    leftMin[0] = a[0];
    rightMin[a.size() - 1] = a.back(); 
        
    for(int i = 1; i < a.size(); ++i)
    {
        leftMin[i] = min(leftMin[i-1], a[i-1]);
        
        rightMin[rightMin.size() - 1 - i] = min(rightMin[rightMin.size() - i], a[rightMin.size() - i]);
    }
    
    for(int i = 1 ; i < a.size() - 1; ++i)
    {
        if (a[i] > leftMin[i] && a[i] > rightMin[i])
            continue;
        
        ++answer;
    }
    
    return answer;
}

File: 프로그래머스/3/68646. 풍선 터트리기/풍선 터트리기.java
public class Solution {

    public static int solution(int[] a) {
        int answer = 0;

        int[] leftMin = new int[a.length];
        int[] rightMin = new int[a.length];

        leftMin[0] = a[0];
        rightMin[a.length - 1] = a[a.length - 1];

        for (int i = 1; i < a.length; ++i) {
            leftMin[i] = Math.min(leftMin[i - 1], a[i - 1]);
        }

        for (int i = a.length - 2; i >= 0; --i) {
            rightMin[i] = Math.min(rightMin[i + 1], a[i + 1]);
        }

        for (int mid = 0; mid < a.length; mid++) {
            int num = a[mid]; // 가능하냐 안하냐;

            if (num > leftMin[mid] && num > rightMin[mid]) {
                continue;
            }

            answer++;
        }

        return answer;
    }
}
